starfish/pipeline/algorithmbase.py

```diff
@@ -1,6 +1,39 @@
+from starfish.imagestack.imagestack import ImageStack
+from starfish.intensity_table.intensity_table import IntensityTable
+from starfish.types import LOG
 
 
-class AlgorithmBase:
+class AlgorithmBaseType(type):
+
+    def __init__(cls, name, bases, namespace):
+        super().__init__(name, bases, namespace)
+        if len(bases) != 0:
+            # this is _not_ AlgorithmBase.  Instead, it's a subclass of AlgorithmBase.
+            cls.run = AlgorithmBaseType.run_with_logging(cls.run)
+
+    @staticmethod
+    def run_with_logging(func):
+        def helper(*args, **kwargs):
+            result = func(*args, **kwargs)
+            # Spot detection returns a tuple or IntensityTable, filtering returns an Imagestack
+            if isinstance(result, ImageStack):
+                result.update_log(args[0])
+            # Spot detection
+            elif isinstance(result, tuple) or isinstance(result, IntensityTable):
```

Comment:
There are some methods that take IntensityTables and return IntensityTables (decoders, segmentation assignment). I don't think those are covered here yet, unless I'm missing something. 

I think this class deserves a docstring to explain how different approaches are treated. Some of the context from the PR explanation could probably go here.

Reply:
I think they'd fall under the 'isinstance(result, IntensityTable)' case, but you're right gotta add that docstring

Reply:
My read here is that if the result is an IntensityTable, but the second arg is not an ImageStack, nothing will happen. By my read, the cases I suggest fall into the "nothing happens" category.

Reply:
ahhhhhh I see what your saying, yes you're right